File: Python/KeyboardRow/KeyboardRow.py
# coding:utf-8
import re


class Solution(object):
    # (42ms)
    def findWords(self, words):
        """
        :type words: List[str]
        :rtype: List[str]
        """
        ans = []
        keyboard = ['qwertyuiop', 'asdfghjkl', 'zxcvbnm']
        for word in words:
            for i in xrange(3):
                if sum([1 for letter in word.lower() if letter in keyboard[i]]) == len(word):
                    ans.append(word)
                    break
        return ans

    # (42ms)
    def findWords2(self, words):
        # 该方法使用的是正则表达式，(?i)表示的是忽略大小写
        return filter(re.compile('(?i)([qwertyuiop]*|[asdfghjkl]*|[zxcvbnm]*)$').match, words)

print Solution().findWords2(["Hello", "Alaska", "Dad", "Peace"])

File: Java/KeyboardRow/KeyboardRow.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.regex.Pattern;

/**
 * (6ms)
 * 这个方法使用的是正则表达式。当然还有效率更高的算法，但是代码的简洁性以及可读性远不如正则表达式。
 * @author Phoenix-Z
 *
 */
public class KeyboardRow {

	public static String[] findWords(String[] words) {
		Pattern pattern = Pattern.compile("^(?i)([qwertyuiop]*|[asdfghjkl]*|[zxcvbnm]*)$");
		List<String> ans = new ArrayList<>();
		for(String word : words) {
			if(pattern.matcher(word).matches()) {
				ans.add(word);
			}
		}
		// 这里将list转换成String数组大方法值得学习记住
		return ans.toArray(new String[ans.size()]);
    }
	
	public static void main(String[] args) {
		String[] words = {"Hello", "Alaska", "Dad", "Peace"};
		System.out.println(Arrays.toString(findWords(words)));
	}
}
